base/utf.jl

```diff
@@ -0,0 +1,994 @@
+""" ->
+=#
+function check_string_utf32(dat::Vector{UInt32}, len::Int, options::Integer=0)
+    local ch::UInt32
+    local totalchar=0, num2byte=0, num3byte=0, num4byte=0, flags::UInt=0
+    local pos = 0
+    @inbounds while pos < len
+        ch = dat[pos += 1]
+        totalchar += 1
+        if ch > 0x7f
+            if ch < 0x100
+                num2byte += 1
+                flags |= UTF_LATIN1
+            elseif ch < 0x800
+                num2byte += 1
+                flags |= UTF_UNICODE2
+            elseif ch > 0xffff
+                (ch > 0x10ffff) && utf_errfunc(UTF_ERR_INVALID, pos, ch)
+                num4byte += 1
+            elseif !is_surrogate_char(ch)
+                num3byte += 1
+            elseif is_surrogate_lead(ch)
+                pos == len && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos, ch)
+                # next character *must* be a trailing surrogate character
+                ch = dat[pos += 1]
+                !is_surrogate_trail(ch) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, ch)
+                num4byte += 1
+                (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos, ch)
+                flags |= UTF_SURROGATE
+            else
+                utf_errfunc(UTF_ERR_NOT_LEAD, pos, ch)
+            end
+        end
+    end
+    totalchar, flags | (num3byte == 0 ? 0 : UTF_UNICODE3) | (num4byte == 0 ? 0 : UTF_UNICODE4), num4byte, num3byte, num2byte
+end
+
+function check_string_abs(str::AbstractString, options::Integer=0)
+    local ch::UInt32
+    local totalchar=0, num2byte=0, num3byte=0, num4byte=0, flags::UInt=0
+    local pos = start(str)
+    local len = endof(str)
+    @inbounds while pos < len
+        ch, pos = next(str, pos)
+        totalchar += 1
+        if ch > 0x7f
+            if ch < 0x100
+                num2byte += 1
+                flags |= UTF_LATIN1
+            elseif ch < 0x800
+                num2byte += 1
+                flags |= UTF_UNICODE2
+            elseif ch > 0xffff
+                (ch > 0x10ffff) && utf_errfunc(UTF_ERR_INVALID, pos, ch)
+                num4byte += 1
+            elseif !is_surrogate_char(ch)
+                num3byte += 1
+            elseif is_surrogate_lead(ch)
+                pos == len && utf_errfunc(UTF_ERR_MISSING_SURROGATE, pos, ch)
+                # next character *must* be a trailing surrogate character
+                ch, pos = next(str, pos)
+                !is_surrogate_trail(ch) && utf_errfunc(UTF_ERR_NOT_TRAIL, pos, ch)
+                num4byte += 1
+                (options & UTF_NO_SURROGATES) != 0 && utf_errfunc(UTF_ERR_SURROGATE, pos, ch)
+                flags |= UTF_SURROGATE
+            else
+                utf_errfunc(UTF_ERR_NOT_LEAD, pos, ch)
+            end
+        end
+    end
+    totalchar, flags | (num3byte == 0 ? 0 : UTF_UNICODE3) | (num4byte == 0 ? 0 : UTF_UNICODE4), num4byte, num3byte, num2byte
+end
+
+# Quickly copy and set trailing \0
+macro return_fast_utf_copy(T1, T2, len, dat)
```

Comment:
OK... I wanted to make sure these were all in-line, and for the utf-8 and utf-16 input/output ones, they update one of the variables and return another...  didn't know how I'd do that and guarantee the code would be as fast (since these are in the inner loops)  Any advice how to do so in Julia, without resorting to a macro (I wish I know how to do the macros better in Julia also, in C/C++, they seem a lot simpler and cleaner... all that $(esc(variable)) drives me crazy...)
The "return_fast_utf_copy" I thought would need to be 2 methods, if done as a function... and then make sure it in-lined...